### ISSUE TITLE
Per-Project Tile Stamps

### ISSUE THREAD

**Issue by eishiya** [mapeditor/tiled#3005]
Currently, Tile Stamps are always global. Since different Projects are likely to need different stamps using different tilesets, it would be useful to have tile stamp collections tied to Projects.

The main thing I'm not certain about: should all tile stamps always be tied to a project, like maps and tilesets are, or should there be project-specific stamps that are loaded in addition to a global set, similar to how extensions are handled?

If stamps are always tied to a project:
- There would have to be a location created to store the stamps within the project. Tiled could create this directory when the first stamp is added, but then the user ends up with an extra directory they didn't explicitly create themselves. That's similar to what Tiled currently does, but that's a lot more acceprable when it's in Tiled's files and not the user's project directory.
- The stamp selection UI would remain as simple as it currently is.

If stamps are handled like extensions:
- The default stamps directory would contain global stamps
- Users can add a second directory in the Project properties, similar to the extension directory.
- The Tile Stamps view should be able to show just global stamps, just project stamps, and all stamps together. Tabs perhaps, like in the Tilesets view? Or checkboxes in a submenu?

Perhaps a third option, similar to tying all stamps to a project, is to simply store the Tile Stamps directory as a Project property? The UI for stamps and setting the directory could stay identical, only the place where the tile stamps directory is stored would change (I have no idea where it currently is, but it's not in the Project or session as far as I can see). It would most likely mean that a user's first project would end up with its stamps in the default directory and not with the rest of their project, though.

**Comment by bjorn** [comment]
> Currently, Tile Stamps are always global. Since different Projects are likely to need different stamps using different tilesets, it would be useful to have tile stamp collections tied to Projects.

Ah, actually, the Tile Stamps is stored in the session since Tiled 1.5, while still defaulting to the global location `Preferences::dataLocation() + QLatin1String("/stamps")`. Would that already resolve this issue?

**Comment by eishiya** [comment]
Whoops, guess I should've researched this more before posting!

That said, I think the fact that I was confused is in itself a bad sign xP I think it would be easier to understand and more consistent if the non-default stamps directory was a Project property that could be viewed/set in Project Properties, just like Object Types are. It *would* mean there would be two different places to change the directory, but I don't think that's a problem.

**Comment by bjorn** [comment]
So, the advantage of the default directory being in the application-specific data location, is that stamps can be saved immediately without the need to set up a directory first. I think that's why I kept it there. If we make the default like the project extensions directory, relative to the project, then I'm not sure if we should still automatically create this directory. What's your opinion on that?

**Comment by eishiya** [comment]
If you mean should `<Tiled application directory>/stamps/` be created, then I think yes. Just as there's a default extensions directory, I think it's good to have a default stamps directory, so that users can get into using Tiled quickly, and can create Project-specific directories if they need to.

**Comment by bjorn** [comment]
No, that directory is already created automatically. I was asking if we move this to the project, and have a project-relative default like "stamps" (as with the "extensions" directory), should we also automatically create this directory?

But I guess the project default should be empty, and in case it's empty, we take the current global directory. Then, when the user wants to set a different stamps directory for the project, the directory will first need to be manually created.

**Comment by eishiya** [comment]
> But I guess the project default should be empty, and in case it's empty, we take the current global directory. Then, when the user wants to set a different stamps directory for the project, the directory will first need to be manually created.

This seems sensible to me. But, rather than checking for the directory being *empty*, I think it should check for the project directory *existing*. I think it's rather different when a directory doesn't exist (the default should be used) from when the directory exists (i.e. was created by the user) and is empty (there are no stamps in it yet).

**Comment by bjorn** [comment]
I meant empty as in "empty string", so the default value for the project is "", and in this case, it will fall back to the global directory. I think if you set up a directory, but this directory does not exist, there should probably be some error message somewhere (either Issues or in the Stamps Dock).

**Comment by eishiya** [comment]
Ah, I see. That sounds good!
